File: src/CattyTests/Bricks/SetLookBrickTests.m
/**
 *  Copyright (C) 2010-2015 The Catrobat Team
 *  (http://developer.catrobat.org/credits)
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  An additional term exception under section 7 of the GNU Affero
 *  General Public License, version 3, is available at
 *  (http://developer.catrobat.org/license_additional_term)
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program.  If not, see http://www.gnu.org/licenses/.
 */

#import <XCTest/XCTest.h>
#import "BrickTests.h"
#import "WhenScript.h"
#import "Pocket_Code-Swift.h"

@interface SetLookBrickTests : BrickTests
@end

@implementation SetLookBrickTests

- (void)setUp
{
    [super setUp];
    // Put setup code here. This method is called before the invocation of each test method in the class.
}

- (void)tearDown
{
    // Put teardown code here. This method is called after the invocation of each test method in the class.
    [super tearDown];
}

- (void)testSetLookBrick
{
//    SpriteObject *object = [[SpriteObject alloc] init];
//    Program *program = [Program defaultProgramWithName:@"a" programID:nil];
//    CBSpriteNode *spriteNode = [[CBSpriteNode alloc] initWithSpriteObject:object];
//    object.spriteNode = spriteNode;
//    CBPlayerScene *scene = [[CBPlayerScene alloc] init];
//    [scene addChild:spriteNode];
//    object.program = program;
//
//    NSBundle *bundle = [NSBundle bundleForClass:[self class]];
//    NSString * filePath = [bundle pathForResource:@"test.png" ofType:nil];
//    NSData *imageData = UIImagePNGRepresentation([UIImage imageWithContentsOfFile:filePath]);
//    Look *look = [[Look alloc] initWithName:@"test" andPath:@"test.png"];
//    [imageData writeToFile:[NSString stringWithFormat:@"%@images/%@", [object projectPath], @"test.png"]atomically:YES];
//    Look *look1 = [[Look alloc] initWithName:@"test2" andPath:@"test2.png"];
//    [imageData writeToFile:[NSString stringWithFormat:@"%@images/%@", [object projectPath], @"test2.png"]atomically:YES];
//
//    Script *script = [[WhenScript alloc] init];
//    script.object = object;
//    NextLookBrick *brick = [[NextLookBrick alloc] init];
//    brick.script = script;
//    [object.lookList addObject:look];
//    [object.lookList addObject:look1];
//    object.spriteNode.currentLook = nil;
//    object.spriteNode.currentUIImageLook = [UIImage imageWithContentsOfFile:filePath];
//    object.spriteNode.currentLookBrightness = 0.0f;
//    dispatch_block_t action = [brick actionBlock];
//    action();
//    XCTAssertEqual(spriteNode.currentLook, look, @"NextLookBrick not correct");
//    [Program removeProgramFromDiskWithProgramName:program.header.programName programID:program.header.programID];
}

- (void)testSetLookBrick2
{
//    SpriteObject* object = [[SpriteObject alloc] init];
//    Program *program = [Program defaultProgramWithName:@"a" programID:nil];
//    CBSpriteNode *spriteNode = [[CBSpriteNode alloc] initWithSpriteObject:object];
//    object.spriteNode = spriteNode;
//    CBPlayerScene *scene = [[CBPlayerScene alloc] init];
//    [scene addChild:spriteNode];
//    object.program = program;
//
//    NSBundle *bundle = [NSBundle bundleForClass:[self class]];
//    NSString *filePath = [bundle pathForResource:@"test.png" ofType:nil];
//    NSData *imageData = UIImagePNGRepresentation([UIImage imageWithContentsOfFile:filePath]);
//    Look* look = [[Look alloc] initWithName:@"test" andPath:@"test.png"];
//    [imageData writeToFile:[NSString stringWithFormat:@"%@images/%@", [object projectPath], @"test.png"]atomically:YES];
//    Look* look1 = [[Look alloc] initWithName:@"test2" andPath:@"test2.png"];
//    [imageData writeToFile:[NSString stringWithFormat:@"%@images/%@", [object projectPath], @"test2.png"]atomically:YES];
//
//    Script *script = [[WhenScript alloc] init];
//    script.object = object;
//
//    NextLookBrick* brick = [[NextLookBrick alloc] init];
//    brick.script = script;
//    [object.lookList addObject:look1];
//    [object.lookList addObject:look];
//    spriteNode.currentLook = nil;
//    spriteNode.currentUIImageLook = [UIImage imageWithContentsOfFile:filePath];
//    spriteNode.currentLookBrightness = 0.0f;
//
//    dispatch_block_t action = [brick actionBlock];
//    action();
//    XCTAssertEqual(spriteNode.currentLook,look1, @"NextLookBrick not correct");
//    [Program removeProgramFromDiskWithProgramName:program.header.programName programID:program.header.programID];
}

@end

File: src/CattyTests/Bricks/SetLookBrickTests.swift
/**
 *  Copyright (C) 2010-2021 The Catrobat Team
 *  (http://developer.catrobat.org/credits)
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  An additional term exception under section 7 of the GNU Affero
 *  General Public License, version 3, is available at
 *  (http://developer.catrobat.org/license_additional_term)
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program.  If not, see http://www.gnu.org/licenses/.
 */

import XCTest

@testable import Pocket_Code

final class SetLookBrickTests: AbstractBrickTest {

    var lookA: Look!
    var lookB: Look!
    var image: UIImage!

    var scene: Scene!
    var object: SpriteObject!
    var spriteNode: CBSpriteNode!
    var script: Script!
    var context: CBScriptContextProtocol!

    override func setUp() {
        super.setUp()

        let filePath = Bundle(for: type(of: self)).path(forResource: "test.png", ofType: nil)!
        image = UIImage(contentsOfFile: filePath)

        lookA = LookMock(name: "LookA", absolutePath: filePath)
        lookB = LookMock(name: "LookB", absolutePath: filePath)
        scene = Scene(name: "scene")

        scene.project = Project()

        object = SpriteObject()
        scene.add(object: object)

        object.lookList.add(lookA!)
        object.lookList.add(lookB!)

        spriteNode = CBSpriteNode(spriteObject: object)
        object.spriteNode = spriteNode

        script = Script()
        script.object = object

        context = CBScriptContext(script: script, spriteNode: spriteNode, formulaInterpreter: formulaInterpreter, touchManager: formulaInterpreter.touchManager)
    }

    func testMutableCopy() {
        let brick = SetLookBrick()
        let look = Look(name: "lookToCopy", filePath: "look")
        brick.look = look

        let copiedBrick: SetLookBrick = brick.mutableCopy(with: CBMutableCopyContext()) as! SetLookBrick

        XCTAssertTrue(brick.isEqual(to: copiedBrick))
        XCTAssertFalse(brick === copiedBrick)
        XCTAssertTrue(brick.look!.isEqual(copiedBrick.look))
        XCTAssertTrue(copiedBrick.look === brick.look)
    }

    func testInstructionWithCache() {
        let imageCacheMock = RuntimeImageCacheMock(imagesOnDisk: [:],
                                                   cachedImages: [CachedImage(path: lookB.path(for: scene)!, image: image)])

        let brick = SetLookBrick()
        brick.script = script
        brick.look = lookB

        spriteNode.currentLook = lookA

        let instruction = brick.actionBlock(imageCache: imageCacheMock)
        instruction()

        XCTAssertEqual(lookB, spriteNode.currentLook)
        XCTAssertEqual(0, imageCacheMock.loadImageFromDiskCalledPaths.count)
    }

    func testInstructionWithoutCache() {
        let imageCacheMock = RuntimeImageCacheMock(imagesOnDisk: [lookB.path(for: scene)!: image],
                                                   cachedImages: [])

        let brick = SetLookBrick()
        brick.script = script
        brick.look = lookB

        spriteNode.currentLook = lookA

        let instruction = brick.actionBlock(imageCache: imageCacheMock)
        instruction()

        XCTAssertEqual(lookB, spriteNode.currentLook)
        XCTAssertEqual(1, imageCacheMock.loadImageFromDiskCalledPaths.count)
        XCTAssertEqual(lookB.path(for: scene), imageCacheMock.loadImageFromDiskCalledPaths.first!)
    }

    func testInstructionWithoutLook() {
        let imageCacheMock = RuntimeImageCacheMock(imagesOnDisk: [lookB.path(for: scene)!: image],
                                                   cachedImages: [])

        let brick = SetLookBrick()
        brick.script = script
        brick.look = nil

        spriteNode.currentLook = lookA

        let instruction = brick.actionBlock(imageCache: imageCacheMock)
        instruction()

        XCTAssertEqual(lookA, spriteNode.currentLook)
        XCTAssertEqual(0, imageCacheMock.loadImageFromDiskCalledPaths.count)
    }

    func testInstructionInvalidImage() {
        let imageCacheMock = RuntimeImageCacheMock(imagesOnDisk: [:], cachedImages: [])
        let lookWithInvalidPath = LookMock(name: "LookWithInvalidPath", absolutePath: "invalidPath")

        let brick = SetLookBrick()
        brick.script = script
        brick.look = lookWithInvalidPath

        spriteNode.currentLook = lookA

        let instruction = brick.actionBlock(imageCache: imageCacheMock)
        instruction()

        XCTAssertEqual(lookA, spriteNode.currentLook)
        XCTAssertEqual(1, imageCacheMock.loadImageFromDiskCalledPaths.count)
        XCTAssertEqual(lookWithInvalidPath.path(for: scene), imageCacheMock.loadImageFromDiskCalledPaths.first!)
    }

    func testIsEqualReturnTrue() {
        let brick1 = SetLookBrick()
        brick1.look = lookA

        let brick2 = SetLookBrick()
        brick2.look = lookA

        XCTAssertTrue(brick1.isEqual(to: brick2))
    }

    func testIsEqualReturnFalse() {
        let brick1 = SetLookBrick()
        brick1.look = lookA

        let brick2 = SetBackgroundBrick()
        brick2.look = lookA

        XCTAssertFalse(brick1.isEqual(to: brick2))
    }
}
